components/script/dom/htmlelement.rs

```diff
@@ -914,45 +916,64 @@ impl HTMLElement {
+        let parent = if self.is_implicit_summary_element() {
+            DomRoot::downcast::<HTMLDetailsElement>(self.containing_shadow_root().unwrap().Host())
+                .unwrap()
+        } else {
+            self.upcast::<Node>()
+                .GetParentNode()
+                .and_then(DomRoot::downcast::<HTMLDetailsElement>)
+                .unwrap()
+        };
 
-        // Step 3
-        parent_details
-            .downcast::<HTMLDetailsElement>()
-            .unwrap()
-            .toggle();
+        // Step 3. If the open attribute is present on parent, then remove it.
+        // Otherwise, set parent's open attribute to the empty string.
+        parent.toggle();
     }
 
-    // https://html.spec.whatwg.org/multipage/#summary-for-its-parent-details
-    fn is_summary_for_its_parent_details(&self) -> bool {
-        // Step 1
-        let summary_node = self.upcast::<Node>();
-        if !summary_node.has_parent() {
-            return false;
+    /// <https://html.spec.whatwg.org/multipage/#summary-for-its-parent-details>
+    fn is_a_summary_for_its_parent_details(&self) -> bool {
+        if self.is_implicit_summary_element() {
+            return true;
         }
 
-        // Step 2
-        let parent = &summary_node.GetParentNode().unwrap();
+        // Step 1. If this summary element has no parent, then return false.
+        // Step 2. Let parent be this summary element's parent.
+        let Some(parent) = self.upcast::<Node>().GetParentNode() else {
+            return false;
+        };
 
-        // Step 3
-        if !parent.is::<HTMLDetailsElement>() {
+        // Step 3. If parent is not a details element, then return false.
+        let Some(details) = parent.downcast::<HTMLDetailsElement>() else {
             return false;
-        }
+        };
 
-        // Step 4 & 5
-        let first_summary_element = parent
-            .child_elements()
-            .find(|el| el.local_name() == &local_name!("summary"));
-        match first_summary_element {
-            Some(first_summary) => &*first_summary == self.as_element(),
-            None => false,
-        }
+        // Step 4. If parent's first summary element child is not this summary
+        // element, then return false.
+        // Step 5. Return true.
+        details
+            .find_corresponding_summary_element()
+            .is_some_and(|summary| &*summary == self.upcast())
+    }
+
+    /// Whether or not this is an implicitly generated `<summary>`
+    /// element for a UA `<details>` shadow tree
+    fn is_implicit_summary_element(&self) -> bool {
+        // Note that non-implicit summary elements are not actually inside
+        // the UA shadow tree, they're only assigned to a slot inside it.
+        // Therefore they don't cause false positives here
+        self.containing_shadow_root()
+            .as_deref()
+            .map(ShadowRoot::Host)
+            .is_some_and(|host| host.is::<HTMLDetailsElement>())
```

Comment:
Hrm. I'm not sure if this is a problem or not but isn't this also true for the `summary` and `descendant` elements of the `ShadowTree`?

Reply:
Ah, right and slot elements are obviously not `HTMLDetailsElement`. Got it. Thanks.